pyg_lib/csrc/ops/cpu/matmul_kernel.cpp

```diff
@@ -1,46 +1,412 @@
+
+        size_t group_idx = 0;
+        for (const auto group_kv : groups) {
+          int m = group_kv.first;
+          const auto indices = group_kv.second;
+
+          ms[group_idx] = m;
+          group_sizes[group_idx] = indices.size();
+          ++group_idx;
+
+          for (const auto offset_idx : indices) {
+            const auto offset = offsets[offset_idx];
+            src0.push_back(src0_base_ptr + offset.src0_offset);
+            src1.push_back(src1_base_ptr + offset.src1_offset);
+            dst.push_back(dst_base_ptr + offset.dst_offset);
+          }
+        }
+
+        auto src0_ptrs = const_cast<const scalar_t**>(src0.data());
+        auto src1_ptrs = const_cast<const scalar_t**>(src1.data());
+        auto dst_ptrs = dst.data();
+
+#if AT_MKL_SEQUENTIAL()
+        // unlikely to happen - requires Torch to be built from source with
+        // explicit flag denoting MKL sequential version
+        parallel_mkl_blas_gemm_batched(ms, ns, ks, alpha, src0_ptrs, ld_src0,
+                                       src1_ptrs, ld_src1, beta, dst_ptrs,
+                                       ld_dst, group_count, group_sizes);
+#else
+        mkl_blas_gemm_batched(ms.data(), ns.data(), ks.data(), alpha.data(),
+                              src0_ptrs, ld_src0.data(), src1_ptrs, ld_src1.data(),
+                              beta.data(), dst_ptrs, ld_dst.data(), group_count,
+                              group_sizes.data());
+#endif
+      });
+}
+
 at::Tensor segment_matmul_kernel(const at::Tensor& input,
                                  const at::Tensor& ptr,
                                  const at::Tensor& other) {
   const auto size = pyg::utils::size_from_ptr(ptr).cpu();
   const auto sizes = at::IntArrayRef(size.data_ptr<int64_t>(), size.numel());
-  const auto out = input.new_empty({input.size(0), other.size(-1)});
-
-  auto outs = out.split_with_sizes(/*split_size=*/sizes, /*dim=*/0);
-  for (auto& out_part : outs) {
-    out_part.resize_(0);
-  }
+  const auto input_contig = input.contiguous();
+  const auto other_contig = other.contiguous();
+  auto out = input_contig.new_empty({input.size(0), other.size(-1)});
 
-  grouped_matmul_out_kernel(
-      input.contiguous().split_with_sizes(/*split_size=*/sizes, /*dim=*/0),
-      other.contiguous().split(/*split_size=*/1, /*dim=*/0), outs);
+  AT_DISPATCH_ALL_TYPES(
+      input_contig.scalar_type(), "segment_matmul_kernel", [&] {
+        c10::guts::if_constexpr<WITH_MKL_BLAS() && AT_MKL_ENABLED() &&
+                                is_blas_library_type<scalar_t>::value>(
+            [&](auto _) {
+              segment_matmul_out_kernel_mkl_impl(input_contig, other_contig,
+                                                 out, sizes);
+            },
+            [&](auto _) {
+              auto outs = out.split_with_sizes(/*split_size=*/sizes, /*dim=*/0);
+              for (auto& out_part : outs) {
+                out_part.resize_(0);
```

Comment:
`resize_(0)` is kind of strange, why do we need it?

Reply:
After splitting the `other` tensor, we get shapes like (1, K, M), so the output is expected to be in shape (1, N, M). That is not happening and a warning is generated. The code here is to fix that behavior (recommended in the warning). On the 2nd look now, it looks weird to me as well, so I changed it to a more intuitive `unsqueeze_`.